# Setup

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
def dataframe_info(df):
    report = pd.DataFrame(columns=['Column', 'Data Type', 'Unique Count', 'Unique Sample', 'Missing Values', 'Missing Percentage'])
    for column in df.columns:
        data_type = df[column].dtype
        unique_count = df[column].nunique()
        unique_sample = df[column].unique()[:5]
        missing_values = df[column].isnull().sum()
        missing_percentage = (missing_values / len(df)) * 100
        report = pd.concat([report, pd.DataFrame({'Column': [column],
                                                      'Data Type': [data_type],
                                                      'Unique Count': [unique_count],
                                                      'Unique Sample': [unique_sample],
                                                      'Missing Values': [missing_values],
                                                      'Missing Percentage': [missing_percentage.round(4)]})],
                             ignore_index=True)
    return report

# Loading the Data

In [3]:
train_data = pd.read_csv('D:\\CDS590\\data\\processed\\train_data.csv')
test_data = pd.read_csv('D:\\CDS590\\data\\processed\\test_data.csv')

In [70]:
dataframe_info(train_data)

C:\Users\ASUS TUF\AppData\Local\Temp\ipykernel_412\3841810920.py:9: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  report = pd.concat([report, pd.DataFrame({'Column': [column],


Column Data Type Unique Count  \
0                route_id     int64            2   
1                distance   float64           15   
2                   month     int64            4   
3         day_of_week_sin   float64            5   
4             minutes_cos   float64          110   
5              avg_temp_C   float64            4   
6             temperature   float64           13   
7               precip_mm   float64           26   
8                humidity   float64           30   
9              visibility   float64            6   
10               pressure   float64           10   
11            cloud_cover   float64           69   
12             wind_speed   float64           10   
13         Speed_Category     int64            2   
14  start_stop_id_encoded     int64            7   
15    end_stop_id_encoded     int64            7   
16               duration   float64          281   

                                        Unique Sample Missing Values  \
0                                            [10, 11]              0   
1   [-0.5619826777423147, -0.0814840550832646, 0.3...              0   
2                                        [3, 2, 5, 4]              0   
3   [-0.433883739117558, 0.9749279121818236, 0.0, ...              0   
4   [0.4848096202463371, 0.8746197071393957, 0.642...              0   
5   [-0.57228042328685, 0.433766690262008, -1.5783...              0   
6   [2.1127378756103314, -0.534445419731945, 0.347...              0   
7   [0.0745795452334384, -0.3089724016813879, 5.70...              0   
8   [-2.476851137319068, 0.6320770499266944, 0.773...              0   
9   [-4.070229021424351, 0.2705098113156833, -0.59...              0   
10  [-1.615354808552546, -0.9701479675098392, 0.32...              0   
11  [1.4561728318080729, 0.9435639107397544, -0.33...              0   
12  [-0.5228408424852269, -1.1291604749470867, 0.6...              0   
13                                             [1, 0]              0   
14                                    [4, 0, 3, 2, 1]              0   
15                                    [5, 1, 4, 3, 6]              0   
16  [1.9833333333333327, 1.8833333333333333, 6.816...              0   

    Missing Percentage  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
5                  0.0  
6                  0.0  
7                  0.0  
8                  0.0  
9                  0.0  
10                 0.0  
11                 0.0  
12                 0.0  
13                 0.0  
14                 0.0  
15                 0.0  
16                 0.0

In [4]:
X_train = train_data.drop('duration', axis=1)
y_train = train_data['duration']
X_test = test_data.drop('duration', axis=1)
y_test = test_data['duration']
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((441, 16), (441,), (111, 16), (111,))

In [58]:
dataframe_info(train_data)

C:\Users\ASUS TUF\AppData\Local\Temp\ipykernel_412\3841810920.py:9: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  report = pd.concat([report, pd.DataFrame({'Column': [column],


Column Data Type Unique Count  \
0                route_id     int64            2   
1                distance   float64           15   
2                   month     int64            4   
3         day_of_week_sin   float64            5   
4             minutes_cos   float64          110   
5              avg_temp_C     int64            4   
6             temperature     int64           13   
7               precip_mm   float64           26   
8                humidity     int64           30   
9              visibility     int64            6   
10               pressure     int64           10   
11            cloud_cover     int64           69   
12             wind_speed     int64           10   
13         Speed_Category     int64            2   
14  start_stop_id_encoded     int64            7   
15    end_stop_id_encoded     int64            7   
16               duration   float64          281   

                                        Unique Sample Missing Values  \
0                                            [10, 11]              0   
1   [1.091282026742783, 1.3487946268033535, 1.5892...              0   
2                                        [3, 2, 5, 4]              0   
3   [-0.433883739117558, 0.9749279121818236, 0.0, ...              0   
4   [0.4848096202463371, 0.8746197071393957, 0.642...              0   
5                                    [23, 24, 22, 25]              0   
6                                [29, 20, 23, 18, 28]              0   
7                           [0.3, 0.0, 4.7, 4.8, 0.1]              0   
8                                [71, 93, 94, 92, 91]              0   
9                                    [5, 10, 9, 7, 3]              0   
10                     [1009, 1010, 1012, 1013, 1011]              0   
11                               [84, 70, 35, 28, 16]              0   
12                                    [2, 1, 4, 5, 3]              0   
13                                             [1, 0]              0   
14                                    [4, 0, 3, 2, 1]              0   
15                                    [5, 1, 4, 3, 6]              0   
16  [1.9833333333333327, 1.8833333333333333, 6.816...              0   

    Missing Percentage  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
5                  0.0  
6                  0.0  
7                  0.0  
8                  0.0  
9                  0.0  
10                 0.0  
11                 0.0  
12                 0.0  
13                 0.0  
14                 0.0  
15                 0.0  
16                 0.0

# Feature Selection

In [5]:
from sklearn.feature_selection import SelectFromModel, SelectKBest, mutual_info_regression, f_regression
from sklearn.linear_model import LassoLarsCV, LassoLarsIC

# Total number of features in the original dataset
all_cols = X_train.columns.to_list()
print(f"All features: {all_cols}\n")

# LassoLarsIC for feature selection
print("LassoLarsIC Feature Selection")
llarsic_selection = SelectFromModel(LassoLarsIC(criterion='aic', max_iter=1000, eps=0.001), max_features=10)
llarsic_selection.fit(X_train, y_train)
llarsic_cols = X_train.columns[llarsic_selection.get_support()].tolist()
print(f"Number of selected features: {len(llarsic_cols)}")
print(f"Selected features: {llarsic_cols}\n")

# SelectKBest with mutual_info_regression for feature selection
print("SelectKBest with mutual_info_regression Feature Selection")
mic_selection = SelectKBest(mutual_info_regression, k=10)
mic_selection.fit(X_train, y_train)
mic_cols = X_train.columns[mic_selection.get_support()].tolist()
print(f"Number of selected features: {len(mic_cols)}")
print(f"Selected features: {mic_cols}\n")

print("SelectKBest with f_regression Feature Selection")
f_regression_selection = SelectKBest(score_func=f_regression, k=10)
f_regression_selection.fit_transform(X_train, y_train) 
f_regression_cols = X_train.columns[f_regression_selection.get_support()].to_list()
print(f"Number of selected features: {len(f_regression_cols)}")
print(f"Selected features: {f_regression_cols}\n")


All features: ['route_id', 'distance', 'month', 'day_of_week_sin', 'minutes_cos', 'avg_temp_C', 'temperature', 'precip_mm', 'humidity', 'visibility', 'pressure', 'cloud_cover', 'wind_speed', 'Speed_Category', 'start_stop_id_encoded', 'end_stop_id_encoded']

LassoLarsIC Feature Selection
Number of selected features: 10
Selected features: ['route_id', 'distance', 'month', 'day_of_week_sin', 'precip_mm', 'humidity', 'pressure', 'Speed_Category', 'start_stop_id_encoded', 'end_stop_id_encoded']

SelectKBest with mutual_info_regression Feature Selection
Number of selected features: 10
Selected features: ['route_id', 'distance', 'minutes_cos', 'temperature', 'precip_mm', 'humidity', 'pressure', 'Speed_Category', 'start_stop_id_encoded', 'end_stop_id_encoded']

SelectKBest with f_regression Feature Selection
Number of selected features: 10
Selected features: ['route_id', 'distance', 'minutes_cos', 'temperature', 'precip_mm', 'pressure', 'cloud_cover', 'Speed_Category', 'start_stop_id_encoded',

# Evaluation 1

In [6]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import PolynomialFeatures
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression, mutual_info_regression
import matplotlib.pyplot as plt

def evaluate_regression_mdl(fitted_model, X_train, X_test, y_train, y_test):
    y_train_pred = fitted_model.predict(X_train)
    y_test_pred = fitted_model.predict(X_test)
    
    return {
        'rmse_train': np.sqrt(mean_squared_error(y_train, y_train_pred)),
        'rmse_test': np.sqrt(mean_squared_error(y_test, y_test_pred)),
        'mae_train': mean_absolute_error(y_train, y_train_pred),
        'mae_test': mean_absolute_error(y_test, y_test_pred),
        'r2_train': r2_score(y_train, y_train_pred),
        'r2_test': r2_score(y_test, y_test_pred)
    }

models = {
    'Linear Regression': LinearRegression(),
    'Ridge': Ridge(alpha=1.0),
    'Polynomial Regression': make_pipeline(PolynomialFeatures(degree=2), Ridge(alpha=1.0)),
    'Decision Tree': DecisionTreeRegressor(max_depth=5, min_samples_leaf=5, random_state=42),
    'Random Forest': RandomForestRegressor(n_estimators=100, max_depth=5, min_samples_leaf=5, random_state=42),
    'XGBoost': GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42),
    'KNN': KNeighborsRegressor(n_neighbors=5),
    'CatBoost': CatBoostRegressor(iterations=100, depth=3, learning_rate=0.1, loss_function='RMSE', verbose=False)
}

feature_selection_methods = {
    'none': None,
    'f_regression': f_regression_cols,
    'mutual_info': mic_cols,
    'llarsic' : llarsic_cols
}

results = []

for model_name, model in models.items():
    for fs_method, selected_features in feature_selection_methods.items():
        print(f"Training {model_name} with {fs_method} feature selection...")
        
        if selected_features is None:
            X_train_fs = X_train
            X_test_fs = X_test
        else:
            X_train_fs = X_train[selected_features]
            X_test_fs = X_test[selected_features]
        
        # Fit model
        model.fit(X_train_fs, y_train)
        
        # Evaluate model
        eval_metrics = evaluate_regression_mdl(model, X_train_fs, X_test_fs, y_train, y_test)
        
        # Perform cross-validation
        cv_scores = cross_val_score(model, X_train_fs, y_train, cv=5, scoring='r2')
        
        # Calculate percentage differences
        rmse_diff_pct = ((eval_metrics['rmse_test'] - eval_metrics['rmse_train']) / eval_metrics['rmse_train']) * 100
        mae_diff_pct = ((eval_metrics['mae_test'] - eval_metrics['mae_train']) / eval_metrics['mae_train']) * 100
        r2_diff_pct = ((eval_metrics['r2_train'] - eval_metrics['r2_test']) / eval_metrics['r2_train']) * 100
        
        results.append({
            'model': f"{model_name}_{fs_method}",
            'fs_method': fs_method,
            'rmse_train': eval_metrics['rmse_train'],
            'rmse_test': eval_metrics['rmse_test'],
            'rmse_diff_pct': rmse_diff_pct,
            'mae_train': eval_metrics['mae_train'],
            'mae_test': eval_metrics['mae_test'],
            'mae_diff_pct': mae_diff_pct,
            'r2_train': eval_metrics['r2_train'],
            'r2_test': eval_metrics['r2_test'],
            'r2_diff_pct': r2_diff_pct,
            'num_feat': X_train_fs.shape[1],
            'cv_r2_mean': cv_scores.mean()
        })

# Convert results to DataFrame
results_df = pd.DataFrame(results)

# Sort results by Test R2 score
results_df = results_df.sort_values('cv_r2_mean', ascending=False).reset_index(drop=True)

# Display results
print("\nModel Comparison Results:")
print(results_df)

# Save results to CSV
results_df.to_csv('model_basicfs_comparison.csv', index=False)

Training Linear Regression with none feature selection...
Training Linear Regression with f_regression feature selection...
Training Linear Regression with mutual_info feature selection...
Training Linear Regression with llarsic feature selection...
Training Ridge with none feature selection...
Training Ridge with f_regression feature selection...
Training Ridge with mutual_info feature selection...
Training Ridge with llarsic feature selection...
Training Polynomial Regression with none feature selection...
Training Polynomial Regression with f_regression feature selection...
Training Polynomial Regression with mutual_info feature selection...
Training Polynomial Regression with llarsic feature selection...
Training Decision Tree with none feature selection...
Training Decision Tree with f_regression feature selection...
Training Decision Tree with mutual_info feature selection...
Training Decision Tree with llarsic feature selection...
Training Random Forest with none feature selecti

In [10]:
results_df.set_index('model').style.\
background_gradient(cmap='bone_r', subset=['mae_train', 'mae_test']).\
background_gradient(cmap='viridis', subset=['r2_train', 'r2_test']).\
background_gradient(cmap='magma', subset=['cv_r2_mean']).\
background_gradient(cmap='gist_heat',  low=0.5, high=3, subset=['mae_diff_pct', 'r2_diff_pct'])

# Hyperparameter Tuning

## Weak model

In [82]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import mean_squared_error, r2_score


# Define the parameter grid for Random Forest
param_grid = {
    'n_estimators': [10, 50, 100],               # Fewer estimators for smaller datasets
    'max_depth': [None, 5, 10, 20],              # Limit the depth of the trees
    'min_samples_split': [5, 7, 10, 12],             # Minimum samples to split an internal node
    'min_samples_leaf': [1, 2, 4],               # Minimum samples to be at a leaf node
    'max_features': ['sqrt', 'log2']     # Number of features to consider at each split
}

# Initialize the RandomForestRegressor
rf = RandomForestRegressor(random_state=42)

# Initialize GridSearchCV
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2, scoring='r2')

# Fit GridSearchCV to the data
grid_search.fit(X_train[f_regression_cols], y_train)

# Get the best parameters and the best model
best_params = grid_search.best_params_
best_rf = grid_search.best_estimator_

print("Best parameters found: ", best_params)

# Evaluate the best model on the test set
y_train_pred = best_rf.predict(X_train[f_regression_cols])
y_test_pred = best_rf.predict(X_test[f_regression_cols])

train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

print("Train RMSE: ", train_rmse)
print("Test RMSE: ", test_rmse)
print("Train R2: ", r2_score(y_train, y_train_pred))
print("Test R2: ", r2_score(y_test, y_test_pred))

# Extract feature importances
feature_importances = pd.Series(best_rf.feature_importances_, index=X_train[f_regression_cols].columns).sort_values(ascending=False)

# Get the list of selected features based on importance
rf_selected_features = feature_importances.index.tolist()

print("Selected Features: ", rf_selected_features)


Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Best parameters found:  {'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 12, 'n_estimators': 10}
Train RMSE:  1.4346646979646016
Test RMSE:  1.4241482054623553
Train R2:  0.90274341370668
Test R2:  0.8807984375593132
Selected Features:  ['distance', 'minutes_cos', 'start_stop_id_encoded', 'Speed_Category', 'route_id', 'humidity', 'end_stop_id_encoded', 'precip_mm', 'pressure', 'temperature']


c:\Users\ASUS TUF\anaconda3\envs\cds590_env\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [83]:
rf_selected_features

['distance',
 'minutes_cos',
 'start_stop_id_encoded',
 'Speed_Category',
 'route_id',
 'humidity',
 'end_stop_id_encoded',
 'precip_mm',
 'pressure',
 'temperature']

## GA

In [11]:
from sklearn_genetic import GAFeatureSelectionCV

ga_rf = GAFeatureSelectionCV(estimator=RandomForestRegressor(random_state=42, max_depth=None, min_samples_split=12, min_samples_leaf=1, n_estimators=10, max_features='sqrt'),
                            cv=3, scoring='neg_root_mean_squared_error', population_size=50,
                            crossover_probability=0.8, mutation_probability=0.1, n_jobs=-1,
                            generations=20, verbose=True, keep_top_k=3,\
                            tournament_size=20, elitism=True, refit=True)
ga_rf.fit(X_train.values, y_train)
ga_rf_cols = np.array(X_train.columns.to_list())[ga_rf.best_features_].tolist()

gen	nevals	fitness	fitness_std	fitness_max	fitness_min
0  	50    	-2.6731	0.588537   	-1.94059   	-4.06886   
1  	94    	-2.0002	0.0461798  	-1.94059   	-2.127     
2  	93    	-1.93011	0.0133545  	-1.90953   	-1.95102   
3  	87    	-1.90978	0.000853958	-1.90953   	-1.91268   
4  	86    	-1.90953	0          	-1.90953   	-1.90953   
5  	88    	-1.90953	0          	-1.90953   	-1.90953   
6  	93    	-1.90953	0          	-1.90953   	-1.90953   
7  	94    	-1.90953	0          	-1.90953   	-1.90953   
8  	90    	-1.90953	0          	-1.90953   	-1.90953   
9  	91    	-1.90953	0          	-1.90953   	-1.90953   
10 	84    	-1.90953	0          	-1.90953   	-1.90953   
11 	90    	-1.90953	0          	-1.90953   	-1.90953   
12 	88    	-1.90953	0          	-1.90953   	-1.90953   
13 	89    	-1.90953	0          	-1.90953   	-1.90953   
14 	90    	-1.90953	0          	-1.90953   	-1.90953   
15 	87    	-1.90953	0          	-1.90953   	-1.90953   
16 	89    	-1.90953	0          	-1.90953   	-1.9095

In [12]:
ga_rf_cols

['route_id',
 'distance',
 'month',
 'temperature',
 'humidity',
 'visibility',
 'pressure',
 'Speed_Category',
 'start_stop_id_encoded',
 'end_stop_id_encoded']

# Eval 2

In [13]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import PolynomialFeatures
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression, mutual_info_regression
import matplotlib.pyplot as plt

def evaluate_regression_mdl(fitted_model, X_train, X_test, y_train, y_test):
    y_train_pred = fitted_model.predict(X_train)
    y_test_pred = fitted_model.predict(X_test)
    
    return {
        'rmse_train': np.sqrt(mean_squared_error(y_train, y_train_pred)),
        'rmse_test': np.sqrt(mean_squared_error(y_test, y_test_pred)),
        'mae_train': mean_absolute_error(y_train, y_train_pred),
        'mae_test': mean_absolute_error(y_test, y_test_pred),
        'r2_train': r2_score(y_train, y_train_pred),
        'r2_test': r2_score(y_test, y_test_pred)
    }

models = {
    'Linear Regression': LinearRegression(),
    'Ridge': Ridge(alpha=1.0),
    'Polynomial Regression': make_pipeline(PolynomialFeatures(degree=2), Ridge(alpha=1.0)),
    'Decision Tree': DecisionTreeRegressor(max_depth=5, min_samples_leaf=5, random_state=42),
    'Random Forest': RandomForestRegressor(n_estimators=100, max_depth=5, min_samples_leaf=5, random_state=42),
    'XGBoost': GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42),
    'KNN': KNeighborsRegressor(n_neighbors=5),
    'CatBoost': CatBoostRegressor(iterations=100, depth=3, learning_rate=0.1, loss_function='RMSE', verbose=False)
}

feature_selection_methods = {
    'none': None,
    'f_regression': f_regression_cols,
    'mutual_info': mic_cols,
    'llarsic' : llarsic_cols,
    'ga': ga_rf_cols
}

results = []

for model_name, model in models.items():
    for fs_method, selected_features in feature_selection_methods.items():
        print(f"Training {model_name} with {fs_method} feature selection...")
        
        if selected_features is None:
            X_train_fs = X_train
            X_test_fs = X_test
        else:
            X_train_fs = X_train[selected_features]
            X_test_fs = X_test[selected_features]
        
        # Fit model
        model.fit(X_train_fs, y_train)
        
        # Evaluate model
        eval_metrics = evaluate_regression_mdl(model, X_train_fs, X_test_fs, y_train, y_test)
        
        # Perform cross-validation
        cv_scores = cross_val_score(model, X_train_fs, y_train, cv=5, scoring='r2')
        
        # Calculate percentage differences
        rmse_diff_pct = ((eval_metrics['rmse_test'] - eval_metrics['rmse_train']) / eval_metrics['rmse_train']) * 100
        mae_diff_pct = ((eval_metrics['mae_test'] - eval_metrics['mae_train']) / eval_metrics['mae_train']) * 100
        r2_diff_pct = ((eval_metrics['r2_train'] - eval_metrics['r2_test']) / eval_metrics['r2_train']) * 100
        
        results.append({
            'model': f"{model_name}_{fs_method}",
            'fs_method': fs_method,
            'rmse_train': eval_metrics['rmse_train'],
            'rmse_test': eval_metrics['rmse_test'],
            'rmse_diff_pct': rmse_diff_pct,
            'mae_train': eval_metrics['mae_train'],
            'mae_test': eval_metrics['mae_test'],
            'mae_diff_pct': mae_diff_pct,
            'r2_train': eval_metrics['r2_train'],
            'r2_test': eval_metrics['r2_test'],
            'r2_diff_pct': r2_diff_pct,
            'num_feat': X_train_fs.shape[1],
            'cv_r2_mean': cv_scores.mean()
        })

# Convert results to DataFrame
results_df = pd.DataFrame(results)

# Sort results by Test R2 score
results_df = results_df.sort_values('cv_r2_mean', ascending=False).reset_index(drop=True)

# Display results
print("\nModel Comparison Results:")
print(results_df)

# Save results to CSV
results_df.to_csv('model_basicfs_comparison.csv', index=False)

Training Linear Regression with none feature selection...
Training Linear Regression with f_regression feature selection...
Training Linear Regression with mutual_info feature selection...
Training Linear Regression with llarsic feature selection...
Training Linear Regression with ga feature selection...
Training Ridge with none feature selection...
Training Ridge with f_regression feature selection...
Training Ridge with mutual_info feature selection...
Training Ridge with llarsic feature selection...
Training Ridge with ga feature selection...
Training Polynomial Regression with none feature selection...
Training Polynomial Regression with f_regression feature selection...
Training Polynomial Regression with mutual_info feature selection...
Training Polynomial Regression with llarsic feature selection...
Training Polynomial Regression with ga feature selection...
Training Decision Tree with none feature selection...
Training Decision Tree with f_regression feature selection...
Traini

In [14]:
results_df.set_index('model').style.\
background_gradient(cmap='bone_r', subset=['mae_train', 'mae_test']).\
background_gradient(cmap='viridis', subset=['r2_train', 'r2_test']).\
background_gradient(cmap='magma', subset=['cv_r2_mean']).\
background_gradient(cmap='gist_heat',  low=0.5, high=3, subset=['mae_diff_pct', 'r2_diff_pct'])

In [17]:
# Get the best parameters
best_params = study.best_params
print('Best parameters:')
for key, value in best_params.items():
    print(f'{key}: {value}')

# If you want to see the best score as well:
print(f'\nBest negative MSE score: {study.best_value}')

Best parameters:
iterations: 691
learning_rate: 0.021930917839362678
depth: 4
l2_leaf_reg: 9.716795056052506
bootstrap_type: Bayesian
random_strength: 9.317429246324037
od_type: Iter
od_wait: 44
bagging_temperature: 0.7078822025613726

Best negative MSE score: 3.648439856331359


In [20]:
import pandas as pd
import numpy as np
from catboost import CatBoostRegressor
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Define the parameter grid for CatBoost
param_grid = {
    'iterations': [500, 1000],
    'learning_rate': [0.01, 0.1],
    'depth': [4, 6, 8],
    'l2_leaf_reg': [1, 3, 5, 7],  # L2 regularization
    'border_count': [32, 128],
    'bagging_temperature': [0, 1],  # Higher values reduce overfitting
    'random_strength': [1, 10],  # Higher values reduce overfitting
    'od_type': ['IncToDec'],  # Use overfitting detector
    'od_wait': [20, 50]  # Number of iterations to wait before stopping
}

# Initialize the CatBoostRegressor
cb = CatBoostRegressor(random_seed=42, verbose=False)

# Initialize GridSearchCV
grid_search = GridSearchCV(estimator=cb, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2, scoring='r2')

# Fit GridSearchCV to the data
grid_search.fit(X_train[f_regression_cols], y_train)

# Get the best parameters and the best model
best_params = grid_search.best_params_
best_cb = grid_search.best_estimator_

print("Best parameters found: ", best_params)

# Evaluate the best model on the train and test sets
y_train_pred = best_cb.predict(X_train[f_regression_cols])
y_test_pred = best_cb.predict(X_test[f_regression_cols])

train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

print("Train RMSE: ", train_rmse)
print("Test RMSE: ", test_rmse)
print("Train R2: ", r2_score(y_train, y_train_pred))
print("Test R2: ", r2_score(y_test, y_test_pred))

# Extract feature importances
feature_importances = pd.Series(best_cb.feature_importances_, index=X_train[f_regression_cols].columns).sort_values(ascending=False)

# Get the list of selected features based on importance
cb_selected_features = feature_importances.index.tolist()

print("Selected Features: ", cb_selected_features)

# Plot learning curves
from catboost import Pool
from matplotlib import pyplot as plt

train_pool = Pool(X_train[f_regression_cols], y_train)
test_pool = Pool(X_test[f_regression_cols], y_test)

best_cb.fit(train_pool, eval_set=test_pool, use_best_model=True, plot=True)

plt.show()

Fitting 5 folds for each of 768 candidates, totalling 3840 fits
Best parameters found:  {'bagging_temperature': 0, 'border_count': 128, 'depth': 4, 'iterations': 1000, 'l2_leaf_reg': 7, 'learning_rate': 0.01, 'od_type': 'IncToDec', 'od_wait': 20, 'random_strength': 10}
Train RMSE:  1.5255052508723075
Test RMSE:  1.5083426708567784
Train R2:  0.890037245153902
Test R2:  0.8662876231591599
Selected Features:  ['distance', 'Speed_Category', 'start_stop_id_encoded', 'minutes_cos', 'route_id', 'end_stop_id_encoded', 'precip_mm', 'temperature', 'cloud_cover', 'pressure']


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

## CatBoost Model Training Process

The CatBoost model underwent a comprehensive training process, as evidenced by the learning curve graph and performance metrics provided. This section details the key aspects of the training:

### Learning Curve Analysis

The graph illustrates the model's learning progression over 1000 iterations:

1. **Convergence**: Both the training (solid line) and test (dotted line) RMSE curves show a steady decrease, indicating consistent learning and improvement in prediction accuracy.

2. **Overfitting Control**: The close proximity of the training and test curves, particularly in later iterations, suggests effective overfitting prevention. This demonstrates CatBoost's built-in regularization mechanisms working as intended.

3. **Early Iterations**: Rapid improvement is observed in the initial stages (0-200 iterations), with the RMSE dropping sharply from around 4.5 to approximately 2.0.

4. **Later Iterations**: The learning rate slows down after about 400 iterations, with both curves showing more gradual improvement. This is typical in gradient boosting models as they fine-tune predictions.

### Performance Metrics

The final model performance metrics are as follows:

- **Training RMSE**: 1.5083426687 (cur)
- **Test RMSE**: 1.63454978740 (test)
- **Best Test RMSE**: 1.50834268700 (best)

These metrics indicate:

1. **Generalization**: The small difference between training and test RMSE (1.5083 vs 1.6345) suggests good generalization capabilities.

2. **Model Selection**: The 'best' test RMSE (1.5083) is lower than the final test RMSE, indicating that the model selection process likely employed early stopping or best model selection to prevent overfitting.

### Training Duration

The training process took 229ms, which is relatively quick and suggests efficient use of computational resources.

### Conclusion

The CatBoost model demonstrates a robust training process with effective learning and generalization. The close alignment of training and test errors, coupled with the gradual convergence of the learning curves, indicates a well-balanced model that avoids overfitting while capturing the underlying patterns in the data. The final RMSE values suggest strong predictive performance, making this model suitable for deployment in production environments.

In [21]:
# Saving the model
best_cb.save_model('best_catboost_model.cbm')


In [22]:
# save the model in JSON format (for version control)
best_cb.save_model('best_catboost_model.json', format='json')

In [11]:
import os
os.environ["PATH"] += os.pathsep + 'C:\\Users\\ASUS TUF\\anaconda3\\pkgs\\python-graphviz-0.20.1-py312haa95532_0\\Lib\\site-packages'

In [2]:
from catboost import CatBoostRegressor
from IPython.display import Image, display
import graphviz

best_cb = CatBoostRegressor()
best_cb.load_model('best_catboost_model.cbm')

# To visualize a specific tree (e.g., tree number 0)
tree_graph = best_cb.plot_tree(
    tree_idx=0,  
    pool=None
)

# Save the graph to a file
graph_file = 'tree_0'
tree_graph.render(graph_file, format='png', cleanup=True)

# Display the graph (in Jupyter Notebook)
display(Image(filename=f'{graph_file}.png'))

# To visualize multiple trees
for i in range(5):  # Visualize first 5 trees
    tree_graph = best_cb.plot_tree(tree_idx=i, pool=None)
    graph_file = f'tree_{i}'
    tree_graph.render(graph_file, format='png', cleanup=True)
    display(Image(filename=f'{graph_file}.png'))

# To get a text representation of a tree
tree_text = best_cb.get_tree_plural(tree_idx=0, output_format='PythonCode')
print(tree_text)

ExecutableNotFound: failed to execute WindowsPath('dot'), make sure the Graphviz executables are on your systems' PATH

In [19]:
from sklearn import inspection
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.model_selection import cross_val_score

cb_mdl = CatBoostRegressor(learning_rate=0.021930917839362678, iterations=691, depth=4,
                           l2_leaf_reg=9.716795056052506, bootstrap_type='Bayesian', random_strength=9.317429246324037,
                           od_type='Iter', od_wait=44, bagging_temperature=0.7078822025613726,
                           random_state=42, verbose=False)
cb_mdl = cb_mdl.fit(X_train[mic_cols], y_train)
cb_feat_imp = cb_mdl.feature_importances_

# Calculate metrics for CatBoost model
cb_y_train_pred = cb_mdl.predict(X_train[mic_cols])
cb_y_test_pred = cb_mdl.predict(X_test[mic_cols])
cb_mae_train = mean_absolute_error(y_train, cb_y_train_pred)
cb_mae_test = mean_absolute_error(y_test, cb_y_test_pred)
cb_r2_train = r2_score(y_train, cb_y_train_pred)
cb_r2_test = r2_score(y_test, cb_y_test_pred)
cb_cv_r2_mean = cross_val_score(cb_mdl, X_train[mic_cols], y_train, cv=5, scoring='r2').mean()

rf_mdl = RandomForestRegressor(random_state=42, max_depth=None, min_samples_split=12, min_samples_leaf=1, n_estimators=10, max_features='sqrt')
rf_mdl = rf_mdl.fit(X_train[mic_cols].to_numpy(), y_train.to_numpy())
rf_feat_imp = rf_mdl.feature_importances_

# Calculate metrics for RandomForest model
rf_y_train_pred = rf_mdl.predict(X_train[mic_cols].to_numpy())
rf_y_test_pred = rf_mdl.predict(X_test[mic_cols].to_numpy())
rf_mae_train = mean_absolute_error(y_train, rf_y_train_pred)
rf_mae_test = mean_absolute_error(y_test, rf_y_test_pred)
rf_r2_train = r2_score(y_train, rf_y_train_pred)
rf_r2_test = r2_score(y_test, rf_y_test_pred)
rf_cv_r2_mean = cross_val_score(rf_mdl, X_train[mic_cols].to_numpy(), y_train.to_numpy(), cv=5, scoring='r2').mean()

# Print metrics
print(f"CatBoost MAE Train: {cb_mae_train}, MAE Test: {cb_mae_test}, R2 Train: {cb_r2_train}, R2 Test: {cb_r2_test}, CV R2 Mean: {cb_cv_r2_mean}")
print(f"RandomForest MAE Train: {rf_mae_train}, MAE Test: {rf_mae_test}, R2 Train: {rf_r2_train}, R2 Test: {rf_r2_test}, CV R2 Mean: {rf_cv_r2_mean}")

# Store cross-validation results
cv_r2_mean = {
    'CatBoost': cb_cv_r2_mean,
    'RandomForest': rf_cv_r2_mean
}

print(f"Cross-Validation R2 Mean: {cv_r2_mean}")


CatBoost MAE Train: 0.8421258624308773, MAE Test: 0.8739314734140842, R2 Train: 0.9113404307330374, R2 Test: 0.8778913274779309, CV R2 Mean: 0.8237007838442535
RandomForest MAE Train: 0.7652782492759457, MAE Test: 0.8368449732404852, R2 Train: 0.90274341370668, R2 Test: 0.8807984375593132, CV R2 Mean: 0.8203412699282179
Cross-Validation R2 Mean: {'CatBoost': 0.8237007838442535, 'RandomForest': 0.8203412699282179}
